### PR TITLE
Telemetry for NuGet only

### PR DESCRIPTION
**Pull request by keithrob**

### CODE REVIEW

File: Tasks/Common/utility-common/package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "utility-common",
-  "version": "1.0.0",
+  "version": "1.0.1",
```
Comment (jahquin-msft) [inline Tasks/Common/utility-common/package.json]: Should be 1.0.2

File: Tasks/Common/utility-common/telemetry.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+                'BUILD_REPOSITORY_PROVIDER': tl.getVariable('BUILD_REPOSITORY_PROVIDER'),
+                'BUILD_REPOSITORY_URI': tl.getVariable('BUILD_REPOSITORY_URI'),
+                'BUILD_SOURCEBRANCH': tl.getVariable('BUILD_SOURCEBRANCH'),
+                'BUILD_SOURCEBRANCHNAME': tl.getVariable('BUILD_SOURCEBRANCHNAME'),
+                'BUILD_SOURCESDIRECTORY': tl.getVariable('BUILD_SOURCESDIRECTORY'),
+                'BUILD_SOURCEVERSION': tl.getVariable('BUILD_SOURCEVERSION'),
+                'BUILD_STAGINGDIRECTORY': tl.getVariable('BUILD_STAGINGDIRECTORY'),
+                'agent.proxyurl': tl.getVariable("agent.proxyurl")
+            });
+            let copy = Object.assign(JSON.parse(commonTelem), telem);
```
Comment (jahquin-msft) [inline Tasks/Common/utility-common/telemetry.ts]: JSON.parse shouldn't be needed if you don't stringify commonTelem first.

File: Tasks/Common/utility-common/telemetry.ts [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+ */
+import * as tl from 'vsts-task-lib/task';
+import * as semver from 'semver';
+
+/**
+ * Utility function to log telemetry.
+ * @param telem A JSON object containing a dictionary of variables that will be appended to
+ * common system vars and loggged.
+ */
+export function emitTelemetry(telem: any) {
```
Comment (jahquin-msft) [inline Tasks/Common/utility-common/telemetry.ts]: Nit. telem -> taskSpecificTelemetry

File: Tasks/Common/nuget-task-common/NuGetToolRunner2.ts [hunk 4]
```diff
@@ -98,14 +99,33 @@ export class NuGetToolRunner2 extends ToolRunner {
...
+        return execResult;
     }
 
     public exec(options?: IExecOptions): Q.Promise<number> {
         options = options || <IExecOptions>{};
         options.env = prepareNuGetExeEnvironment(options.env || process.env, this.settings, this.authInfo);
         return super.exec(options);
     }
+
+    private _logExecResults(exitCode: number, stderr: string) {
```
Comment (jahquin-msft) [inline Tasks/Common/nuget-task-common/NuGetToolRunner2.ts]: This should also be placed in the common telemetry library.

File: Tasks/NuGetCommand/nugetcommandmain.ts [hunk 5]
```diff
@@ -52,4 +55,81 @@ async function main(): Promise<void> {
...
+                'command': tl.getInput('command'),
+                'arguments': tl.getInput('arguments'),
+                'NUGET_EXE_TOOL_PATH_ENV_VAR': tl.getVariable(nuGetGetter.NUGET_EXE_TOOL_PATH_ENV_VAR),
+                'NUGET_EXE_CUSTOM_LOCATION': tl.getVariable(NUGET_EXE_CUSTOM_LOCATION),
+                'searchPatternPack': tl.getPathInput('searchPatternPack'),
+                'configurationToPack': tl.getInput('configurationToPack'),
+                'versioningScheme': tl.getInput('versioningScheme'),
+                'includeReferencedProjects': tl.getBoolInput('includeReferencedProjects'),
+                'versionEnvVar': tl.getInput('versionEnvVar') ?
+                    tl.getInput(tl.getInput('versionEnvVar')) : null,
```
Comment (jahquin-msft) [inline Tasks/NuGetCommand/nugetcommandmain.ts]: This looks weird, why is this needed?

File: Tasks/NuGetCommand/nugetcommandmain.ts [hunk 6]
```diff
@@ -52,4 +55,81 @@ async function main(): Promise<void> {
...
+                'requestedMajorVersion': tl.getInput('requestedMajorVersion'),
```
Comment (jahquin-msft) [inline Tasks/NuGetCommand/nugetcommandmain.ts]: Several of these are repeated, please clean them up.

File: Tasks/NuGetCommand/nugetcommandmain.ts [hunk 7]
```diff
@@ -52,4 +55,81 @@ async function main(): Promise<void> {
...
+                'searchpatternpush': tl.getInput('searchpatternpush'),
+                'selectorconfig': tl.getInput('selectorconfig'),
+                'solution': tl.getInput('solution'),
+                'versioningscheme': tl.getInput('versioningscheme'),
+                'verbositypack': tl.getInput('verbositypack'),
+                'verbositypush': tl.getInput('verbositypush'),
+                'verbosityrestore': tl.getInput('verbosityrestore')
+            });
+
+        let nugetTelem = JSON.parse(nugetTelemStr);
```
Comment (jahquin-msft) [inline Tasks/NuGetCommand/nugetcommandmain.ts]: Same as other comment, parse is not needed if not stringify first.

File: Tasks/NuGetCommand/nugetcommandmain.ts [hunk 8]
```diff
@@ -52,4 +55,81 @@ async function main(): Promise<void> {
...
+                    feedName: tl.getEndpointUrl(id, true).replace(/\W/g, ''),
+                    feedUri: tl.getEndpointUrl(id, true)
+                }
+                ary.push(te);
+                return ary;
+            }, []);
+        }
+        let nugetTelemStr = JSON.stringify(
+            {
+                'command': tl.getInput('command'),
```
Comment (jahquin-msft) [inline Tasks/NuGetCommand/nugetcommandmain.ts]: Curious, what was the reasoning behind a one for all telemetry approach instead of logging it depending on the scenario? I.E. why is this here and not in every command?

Comment (keithrob) [reply Tasks/NuGetCommand/nugetcommandmain.ts]: Simplicity.

File: Tasks/Common/utility-common/package.json (hunk 1 above)
Comment (keithrob) [reply Tasks/Common/utility-common/package.json]: I think you are confusing this package which lives in , utility-common, with bumped versions in nuget-task-common.  utility-common moved from 1.0.0 to 1.0.1 and nuget-task-common moved from 1.0.1 to 1.0.2

Comment (jahquin-msft) [reply Tasks/Common/utility-common/package.json]: Ah yes you're right. And now that you mention it, the telemetry module needs to be compiled with the task, this is done through the task's make.json file.

File: Tasks/NuGetCommand/nugetcommandmain.ts (hunk 6 above)
Comment (keithrob) [reply Tasks/NuGetCommand/nugetcommandmain.ts]: Fixed

File: Tasks/Common/utility-common/package.json (hunk 1 above)
Comment (keithrob) [reply Tasks/Common/utility-common/package.json]: Moot point now.  The common module dependency no longer exists.

File: Tasks/NuGetCommand/nugetcommandmain.ts (hunk 5 above)
Comment (keithrob) [reply Tasks/NuGetCommand/nugetcommandmain.ts]: Fixed.  Should be getVariable(getInput).  This is necessary when a user supplies an input variable containing the name of the environment variable with the actual version.

File: Tasks/NuGetCommand/nugetcommandmain.ts (hunk 7 above)
Comment (keithrob) [reply Tasks/NuGetCommand/nugetcommandmain.ts]: Fixed

File: Tasks/Common/utility-common/telemetry.ts (hunk 2 above)
Comment (keithrob) [reply Tasks/Common/utility-common/telemetry.ts]: Fixed
